S3.php

```diff
@@ -1907,6 +1977,112 @@ private static function __getHash($string)
 		(str_repeat(chr(0x36), 64))) . $string)))));
 	}
 
+
+	/**
+	* Generate the headers for AWS Signature V4
+	* @internal Used by S3Request::getResponse()
+	* @param array $aheaders amzHeaders 
+	* @param array $headers
+	* @param string $method 
+	* @param string $uri
+	* @param string $data
+	* @return array $headers
+	*/
+	public static function __getSignatureV4($aHeaders, $headers, $method='GET', $uri='', $data = '')
+	{		
+		$service = 's3';
+		$region = S3::getRegion();
+		
+		$algorithm = 'AWS4-HMAC-SHA256';
+		$amzHeaders = array();
+		$amzRequests = array();
+		
+		$amzDate =  gmdate( 'Ymd\THis\Z' );
+		$amzDateStamp = gmdate( 'Ymd' );
+
+		// amz-date ISO8601 format ? for aws request		
+		$amzHeaders['x-amz-date'] = $amzDate;
+
+		// CanonicalHeaders
+		foreach ( $headers as $k => $v ) {
+			$amzHeaders[ strtolower( $k ) ] = trim( $v );
+		} 
+		foreach ( $aHeaders as $k => $v ) {
+			$amzHeaders[ strtolower( $k ) ] = trim( $v );
+		} 
+		uksort( $amzHeaders, 'strcmp' );
+
+		// payload
+		$payloadHash = isset($amzHeaders['x-amz-content-sha256']) ? $amzHeaders['x-amz-content-sha256'] :  hash('sha256', $data);
+
+		// parameters
+		$parameters = array();
+		if (strpos($uri, '?')) {
+			list ($uri, $query_str) = @explode("?", $uri);
+			parse_str($query_str, $parameters);
+		}
+
+		// CanonicalRequests
+		$amzRequests[] = $method;
+		$uriQmPos = strpos($uri, '?'); 
+		$amzRequests[] = ($uriQmPos === false ? $uri : substr($uri, 0, $uriQmPos));
+		$amzRequests[] = http_build_query($parameters);
```

Comment:
Better use this: 
`$amzRequests[] = http_build_query($parameters, null, '&', PHP_QUERY_RFC3986);`

Otherwise, the encoding happens with + instead if %20 for spaces and the signature is invalid.